internal/engine/engine.go

```diff
@@ -277,23 +278,49 @@ func (engine *Engine) doPlanResources(ctx context.Context, input *enginev1.PlanR
 
 		maps.Copy(auditTrail.EffectivePolicies, policy.GetMeta().GetSourceAttributes())
 	}
-
-	// get the resource policy check
-	rpName, rpVersion, rpScope := engine.policyAttr(input.Resource.Kind, input.Resource.PolicyVersion, input.Resource.Scope, opts.evalParams)
-	policySet, err = engine.getResourcePolicySet(ctx, rpName, rpVersion, rpScope, opts.LenientScopeSearch())
+	rpEvaluator, err := engine.getRolePolicyEvaluator(ctx, opts.evalParams, ppScope, input.Principal.Roles)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get check for [%s.%s]: %w", rpName, rpVersion, err)
+		return nil, nil, fmt.Errorf("failed to get role policy evaluator: %w", err)
 	}
-
-	if policy := policySet.GetResourcePolicy(); policy != nil {
-		policyEvaluator := planner.ResourcePolicyEvaluator{Policy: policy, Globals: opts.Globals(), SchemaMgr: engine.schemaMgr, NowFn: opts.NowFunc()}
-		plan, err := policyEvaluator.EvaluateResourcesQueryPlan(ctx, input)
+	effectiveRoles := engineinternal.ToSet(input.Principal.Roles)
+	if rpEvaluator != nil {
+		tctx := tracer.Start(opts.tracerSink)
+		evalResult, roles, err := PlannerEvaluateRolePolicy(ctx, tctx, rpEvaluator, input)
 		if err != nil {
 			return nil, nil, err
 		}
+		effInfo, ok := evalResult.Effects[input.Action]
+		if !ok {
+			return nil, nil, errors.New("role policy evaluator unexpected result")
+		}
+		effect := effInfo.Effect
+		if effect != effectv1.Effect_EFFECT_ALLOW {
+			engineinternal.SubstractSets(effectiveRoles, roles)
+		}
+		if result.Empty() {
+			result = mkUnconditionalPolicyPlanResult(effInfo.Scope, effect)
+		}
```

Comment:
I think this needs to be:

```suggestion
		// `effect` can only be ALLOW if the policy has the resource:action pair and has scope permissions set to "override parent".
		// It can only be DENY if the resource:action pair does not exist.
		// In both these cases, the evaluation (for the given role) can terminate within the role policies.
		if effect != effectv1.Effect_EFFECT_NO_MATCH {
			engineinternal.SubstractSets(unresolvedRoles, roles)
			if result.Empty() {
				result = mkUnconditionalPolicyPlanResult(effInfo.Scope, effect)
			}
		}
```